Tests/NetworkingTests/DownloadAPIManagerTests.swift

```diff
@@ -0,0 +1,59 @@
+//
+//  File.swift
+//
+//
+//  Created by Matej Molnár on 01.01.2024.
+//
+
+@testable import Networking
+import XCTest
+
+@NetworkingActor
+final class DownloadAPIManagerTests: XCTestCase {
+    enum DownloadRouter: Requestable {
+        case download(url: URL)
+
+        var baseURL: URL {
+            switch self {
+            case let .download(url):
+                url
+            }
+        }
+
+        var path: String {
+            switch self {
+            case .download:
+                ""
+            }
+        }
+    }
+    
+    func testMultiThreadRequests() async throws {
```

Comment:
This test actually freezes the tests for me when run repeatedly and it just never finishes 😢 

![CleanShot 2024-01-02 at 17 04 02@2x](https://github.com/strvcom/ios-networking/assets/1043677/d9dd1d55-5308-48fb-b9e3-e0ae3ba0a09b)

Reply:
Interestingly the Upload manager test seems to be working fine.

Reply:
Good catch, for some reason it worked for me before, but after I ran clean build folder I am encountering the same issue 😬 Will fix.

Reply:
@johnkodes I've updated the tests, can you try again? 🙏

Reply:
Instead of rewriting it (and probably hiding the issue by the Task) shouldn't we rather find the cause of this? 

When you comment out the `await updateTasks()` then everything works.

Reply:
I agree, but at first I wanted to implement the tests in a way that would prevent them from getting stuck, but after I did that the download test started passing again 😅, that's why I asked you to try it out one more time as well.

Reply:
With the changes with using the `Task` yes, it works, but I don't think that's the correct way, because I have a feeling that we just hid the actual issue 😄 .

Reply:
The problem is that I am not able to debug it since even when I return to the original implementation the tests just pass without freezing, so I am not sure what to do now 😅🤷‍♂️

Reply:
Let me give that a few more tries 😅

Reply:
![CleanShot 2024-01-05 at 07 56 01@2x](https://github.com/strvcom/ios-networking/assets/1043677/8eee4cca-6c3c-4030-93f1-9dd010962848)

No matter how many times I try 😄 and how many clean builds I do 🤷

Reply:
I managed to run 20 000 tests successful tests with changes that I did. 

![CleanShot 2024-01-05 at 14 19 52@2x](https://github.com/strvcom/ios-networking/assets/1043677/194b46d8-ee57-4b15-a7ad-373d16467223)

So I think the problem (I'm definitely not sure) is that the dictionaries you have right now are not thread-safe and those are causing the deadlock/crash. 

What I did was create an actor that wraps a dictionary (with getValue/setValue functions, not a subscript). Mainly for the `downloadStateDict`. The other one seems to not cause any issues but theoretically could as well. 

The way I got to this finding was by removing the `Task` in `updateTasks` which uncovered a crash when accessing the `downloadStateDict`, either in `for await` or when setting the value inside the `.sink`

![CleanShot 2024-01-05 at 14 16 44@2x](https://github.com/strvcom/ios-networking/assets/1043677/d1a50a61-2fb0-47f3-958d-0b23422a367b)

**Here is my assumption of the thread-safe issue and edits of the dictionary:**
![CleanShot 2024-01-05 at 14 35 10@2x](https://github.com/strvcom/ios-networking/assets/1043677/55856fbb-e01c-4476-b18e-0ce6d97c1a12)

Reply:
I am quite baffled by this one 🧐. The dictionary shouldn't be causing problems since all the mutations are being handled on the same actor. So even if the mutating operations are being done on different threads it should be ok, because the actor is supposed to ensure that the data is not being accessed by multiple threads at the same time.

Reply:
And how is the Actor enforced?

Reply:
And are you sure that it means that the whole thing behaves as an actor? (I'm not, so I'm rather asking 😄 ) Eg. it will protect its state? 

Also we are combining Concurrency + Combine, so I wonder whether that breaks everything.

Reply:
I am pretty sure, also I would expect 'Sendable' concurrency warnings on the var properties if that was not the case. What's super weird is that I am not able to replicate your issues.

Reply:
With reverted `DownloadAPIManagerTests` back to using `async` without the Task? 

Xcode 15.1

Reply:
And regarding the other thing, are you getting the crashes? 
Just by removing this `Task`

![CleanShot 2024-01-05 at 16 34 57](https://github.com/strvcom/ios-networking/assets/1043677/eadd8112-5ba4-45c1-a8c5-c2b4bf35a34e)

Reply:
Then running the tests many times

Reply:
@johnkodes I was trying to debug this and the tests are freezing for me even with the `ThreadSafeDictionary` 😬 Could you please share your working solution in some branch? 🙏

Reply:
Shared via Slack

Reply:
@johnkodes @cejanen Ok I think I finally have a solution that works even without using the `ThreadSafeDictionary`🤞. Thanks to @gajddo00 for pointing out that the cause of the issue probably rests in the `asyncResponse()` extension method 🙏. I rewrote the method + removed unnecessary calls to `updateTasks()` function and seems to be working fine, I ran 100 000 test iterations without issues. Let me know what you think.

Reply:
I think that the thread was getting stuck because I first called `resume()` on the `URLSessionTasku` and only afterwards created `sink` for observing the `response`, hence sometimes the `sink` block would never execute and that probably led to the other issues we were experiencing.